### DISCUSSION TITLE
Confused about checks on assignments in collections.

### DISCUSSION THREAD

**Discussion by matthewpersico** [Perl-Oshun/oshun#36] [General]
At [Applying a check to an array variable](https://gist.github.com/thoughtstream/08b7fd48b09c99ae47d6d9f82b913986#applying-a-check-to-an-array-variable), it is stated (**emphasis** mine):

If a check is specified on an array, that check is applied to **each** element of the array, every time **any** element is modified in any way (assigned to, concatenated to, incremented, deleted, etc.)

At [Applying a check to a hash variable](https://gist.github.com/thoughtstream/08b7fd48b09c99ae47d6d9f82b913986#applying-a-check-to-a-hash-variable), it is stated  (**emphasis** mine):

If a check is specified on a hash, that check is applied to the value of each entry in the hash, **[only (sic)] whenever those values are assigned**.

I am confused. Why would you check all ten elements of an array if you are only assigning specific ones, yet check only the newly assigned elements in a hash? Am I reading the statements incorrectly? That array handling seems to be an awful bottleneck if I am going to recheck every element on each `push`, for example.

**Comment by tobyink** [comment]
I imagine that it's written that way with the aim of being able to cover requirements like: a check that the array is sorted numerically. Then you need to check not just that the incoming value is a number, but also that it's been inserted into the correct position in the array, so the entire array needs checking.

However, despite making reference to an ability for user-defined checks to be able to handle whole-array or whole-hash data checks like this, the spec doesn't really provide a way to do it.

**Reply by Ovid** [reply]
@tobyink sorry, I've been busy this week and haven't gotten back to this.  Can you explain what you mean by "the spec doesn't really provide a way to do it"? [Contextual information for check blocks](https://github.com/Perl-Oshun/oshun/wiki/Initial-Specification#contextual-information-for-check-blocks) covers this, unless I've missed your meaning.

Here's a bijective hash: 

```perl
# The values of the hash must remain unique...
check Bijective  :on(HASH)  ($new_value, %target)  {
    return $target{old} eq $new_value
           || not grep { $_ eq $new_value} } values $target{var}->%*;
}
```

**Reply by tobyink** [reply]
You are right, that is covered. `my %numbers :of(NUM)` only constrains each value in the hash, but yeah, with the contextual information it should be possible to add constraints for the whole hash.

**Comment by zmughal** [comment]
@happy-barney, is this the kind of issue you were talking about re: performance [here](https://github.com/Perl-Oshun/oshun/issues/1#issuecomment-1573467987)? That the values (rather than just the variables) should have some idea of what checks they pass?

**Reply by happy-barney** [reply]
it is related by my primary concern was about passing unmodified values here and there:
```
my $foo :contract(x) = $value;
my $bar :contract(x) = $foo;
my $baz :contract(x) = $baz;
```

this problem will benefit from that solution but detailed specification will be needed

**Comment by Ovid** [comment]
@matthewpersico Sorry, that section of the spec needs to be reworded.

> If a check is specified on an array, that check is applied to each new or modified element of the array, every time any element is modified in any way (assigned to, concatenated to, incremented, deleted, etc.)

It's not "all elements", it's only the elements being changed.

**Comment by Ovid** [comment]
@matthewpersico Also, the spec is now at [the wiki](https://github.com/Perl-Oshun/oshun/wiki/Initial-Specification#contextual-information-for-check-blocks) where it can be updated. I'll make the change for this. That gist is now there for historical reasons.

**Comment by Ovid** [comment]
Are we good? Can I close this discussion now, or is there more work needed here?
